documentation/specs/CodeBlock.md

```diff
@@ -0,0 +1,204 @@
+# `CodeBlock` Component Specification
+
+## Overview
+
+A Code Block is used to display readable blocks of code.
+
+### Use Cases
+
+- Use a Code Block to improve readability of embedded code samples with syntax highlighting and automatic line numbering.
+
+### Features
+
+- Supports optional line numbering
+- Supports maximum number of lines before scroll
+- Supports multiple languages
+
+### Prior Art
+
+- [Paste `<CodeBlock />`](https://paste.twilio.design/components/code-block)
+- [Atlassian `<CodeBlock />`](https://atlassian.design/components/code/code-block)
+- [Sentry `<CodeSnippet />`](https://storybook.sentry.dev/?path=/story/components-code-snippet--js)
+
+---
+
+## Design
+
+`CodeBlock` at a base level is pretty simple. It requires a `code` snippet and `language` for highlighting. It can optionally show line numbers and render a maximum number of lines before scrolling. It renders inside of a `Card`.
+
+`CodeBlock` will utilize the `PrismLight` export of `react-syntax-highlighter` to enable syntax highlighting. A lightweight wrapper component called `SyntaxHighlighter` will be created to wrap some details specific to EasyPost, including importing required languages and building some utilities around theme customization.
+
+`CodeBlock` in the future can be expanded to support more complex use cases such as including a header with label, a copy button, etc. In such case, it may make sense to augment `CodeBlock` to support a dot notation paradigm:
+
+```tsx
+<CodeBlock.Container>
+  <CodeBlock.Header label="filename.js" />
+  <CodeBlock.Snippet code={`console.log("Hello world");`} />
+</CodeBlock.Container>
+```
+
+### API
+
+```ts
+export type SnippetLanguages =
+  | "csharp"
+  | "go"
+  | "java"
+  | "javascript"
+  | "json"
+  | "php"
+  | "python"
+  | "ruby"
+  | "shell";
+
+type CodeBlockProps = Partial<Omit<HTMLDivElement, "children">> & {
+  /**
+   * `children` is not supported. Use `code` instead.
+   */
+  children?: never;
+
+  /**
+   * The code snippet to be rendered.
+   */
+  code: string;
+
+  /**
+   * The language of the code snippet.
+   */
+  language: SnippetLanguages;
+
+  /**
+   * Constrains the height of code block to a set number of lines.
+   */
+  maxLines?: number;
+
+  /**
+   * Include line numbers in code block.
+   */
+  showLineNumbers?: boolean;
```

Comment:
i asked design regarding the alternative backbone style that's in figma (the carets) and we agreed to not worry about that for now since the syntax highlighting libs didn't offer a solution for that out of the box